File: backend/app/routers/query.py
"""Query Router"""
from fastapi import APIRouter, HTTPException
from ..schemas.query import (
    QueryRequest, 
    QueryResponse, 
    MaterializedTableCreate,
    ModelInfo,
    JobInfo,
    KnowledgeBaseInfo,
    MindsDBStatus
)
from ..services.mindsdb_service import mindsdb_service
from typing import List

router = APIRouter(prefix="/query", tags=["Query"])


@router.post("", response_model=QueryResponse)
async def execute_query(request: QueryRequest):
    """Execute SQL query on MindsDB"""
    result = await mindsdb_service.execute_query(request.query)
    return QueryResponse(**result)


@router.get("/status", response_model=MindsDBStatus)
async def get_status():
    """Check MindsDB server status"""
    connected, version, error = await mindsdb_service.check_connection()
    return MindsDBStatus(connected=connected, version=version, error=error)


@router.post("/materialized-table")
async def create_materialized_table(request: MaterializedTableCreate):
    """Create a materialized table from source data"""
    result = await mindsdb_service.create_materialized_table(
        table_name=request.table_name,
        source_database=request.source_database,
        source_table=request.source_table,
        columns=request.columns,
        where_clause=request.where_clause,
        limit=request.limit
    )
    
    if result["type"] == "error":
        raise HTTPException(status_code=400, detail=result["error"])
    
    return {"message": f"Materialized table '{request.table_name}' created successfully"}


@router.get("/models", response_model=List[ModelInfo])
async def get_models():
    """Get list of ML models"""
    models = await mindsdb_service.get_models()
    return models


@router.get("/jobs", response_model=List[JobInfo])
async def get_jobs():
    """Get list of scheduled jobs"""
    jobs = await mindsdb_service.get_jobs()
    return jobs


@router.get("/knowledge-bases", response_model=List[KnowledgeBaseInfo])
async def get_knowledge_bases():
    """Get list of knowledge bases"""
    kbs = await mindsdb_service.get_knowledge_bases()
    return kbs

File: frontend/src/stores/query.ts
import { defineStore } from 'pinia';
import { ref, computed } from 'vue';
import { queryApi } from '../api/client';
import type { QueryResult, Model, Job, KnowledgeBase, MindsDBStatus } from '../types';

export const useQueryStore = defineStore('query', () => {
  // State
  const currentQuery = ref('');
  const queryResult = ref<QueryResult | null>(null);
  const queryHistory = ref<{ query: string; timestamp: Date; success: boolean }[]>([]);
  const models = ref<Model[]>([]);
  const jobs = ref<Job[]>([]);
  const knowledgeBases = ref<KnowledgeBase[]>([]);
  const status = ref<MindsDBStatus | null>(null);
  const loading = ref(false);
  const error = ref<string | null>(null);

  // Getters
  const isConnected = computed(() => status.value?.connected ?? false);
  const lastQueries = computed(() => queryHistory.value.slice(-10).reverse());

  // Actions
  async function checkStatus() {
    try {
      status.value = await queryApi.getStatus();
    } catch (e: any) {
      status.value = { connected: false, error: e.message };
    }
  }

  async function executeQuery(query: string) {
    loading.value = true;
    error.value = null;
    currentQuery.value = query;
    
    try {
      queryResult.value = await queryApi.execute(query);
      queryHistory.value.push({
        query,
        timestamp: new Date(),
        success: queryResult.value.type !== 'error'
      });
      
      if (queryResult.value.type === 'error') {
        error.value = queryResult.value.error || 'Query execution failed';
      }
      
      return queryResult.value;
    } catch (e: any) {
      error.value = e.response?.data?.detail || e.message || 'Failed to execute query';
      queryResult.value = {
        type: 'error',
        columns: [],
        data: [],
        row_count: 0,
        error: error.value || undefined
      };
      return queryResult.value;
    } finally {
      loading.value = false;
    }
  }

  async function createMaterializedTable(
    tableName: string,
    sourceDatabase: string,
    sourceTable: string,
    columns: string[] = ['*'],
    whereClause?: string,
    limit?: number
  ) {
    loading.value = true;
    error.value = null;
    
    try {
      await queryApi.createMaterializedTable({
        table_name: tableName,
        source_database: sourceDatabase,
        source_table: sourceTable,
        columns,
        where_clause: whereClause,
        limit
      });
      return true;
    } catch (e: any) {
      error.value = e.response?.data?.detail || e.message || 'Failed to create materialized table';
      return false;
    } finally {
      loading.value = false;
    }
  }

  async function fetchModels() {
    try {
      models.value = await queryApi.getModels();
    } catch (e: any) {
      error.value = e.message || 'Failed to fetch models';
    }
  }

  async function fetchJobs() {
    try {
      jobs.value = await queryApi.getJobs();
    } catch (e: any) {
      error.value = e.message || 'Failed to fetch jobs';
    }
  }

  async function fetchKnowledgeBases() {
    try {
      knowledgeBases.value = await queryApi.getKnowledgeBases();
    } catch (e: any) {
      error.value = e.message || 'Failed to fetch knowledge bases';
    }
  }

  function clearResult() {
    queryResult.value = null;
    error.value = null;
  }

  function clearError() {
    error.value = null;
  }

  function setQuery(query: string) {
    currentQuery.value = query;
  }

  return {
    // State
    currentQuery,
    queryResult,
    queryHistory,
    models,
    jobs,
    knowledgeBases,
    status,
    loading,
    error,
    // Getters
    isConnected,
    lastQueries,
    // Actions
    checkStatus,
    executeQuery,
    createMaterializedTable,
    fetchModels,
    fetchJobs,
    fetchKnowledgeBases,
    clearResult,
    clearError,
    setQuery,
  };
});
